Core/Core/Dashboard/K5/ViewModel/Grades/K5GradesViewModel.swift

```diff
@@ -16,13 +16,112 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 //
 
-class K5GradesViewModel {
+public class K5GradesViewModel: ObservableObject {
 
+    @Published public private(set) var grades: [K5GradeCellViewModel] = []
+    @Published public private(set) var gradingPeriods: [K5GradingPeriod] = []
+    @Published public private(set) var currentGradingPeriod: K5GradingPeriod
+
+    private let env = AppEnvironment.shared
+    private var studentID = ""
+    private lazy var courses = env.subscribe(GetUserCourses(userID: studentID)) { [weak self] in
+        self?.coursesUpdated()
+    }
+    private let defaultCurrentGradingPeriod = K5GradingPeriod(periodID: nil, title: "Current Grading Period")
+
+    // MARK: Refresh
+    private var refreshCompletion: (() -> Void)?
+    private var forceRefresh = false
+
+    init() {
+        studentID = env.currentSession?.userID ?? ""
+        currentGradingPeriod = defaultCurrentGradingPeriod
+        courses.refresh()
+    }
+
+    private func coursesUpdated() {
+        grades.removeAll()
+        gradingPeriods = [defaultCurrentGradingPeriod]
+        grades = courses.filter({ !$0.isHomeroomCourse }).map {
+            return K5GradeCellViewModel(a11yId: "K5GradeCell.\($0.id)",
+                                        title: $0.name ?? "",
+                                        imageURL: $0.imageDownloadURL,
+                                        grade: $0.enrollments?.first?.computedCurrentGrade,
+                                        score: $0.enrollments?.first?.computedCurrentScore,
+                                        color: $0.color,
+                                        courseID: $0.id)
+        }
+        var gradingPeriodModels = courses.compactMap { $0.gradingPeriods }.flatMap { $0 }
+        gradingPeriodModels.sort(by: {
+            guard let date0 = $0.startDate, let date1 = $1.startDate else { return false }
+            return date0 < date1
+        })
+        gradingPeriods.append(contentsOf: gradingPeriodModels.map { return K5GradingPeriod(periodID: $0.id, title: $0.title) })
+        finishRefresh()
+    }
+
+    private func updateEnrollments(for gradingPeriodID: String) {
+        grades.removeAll()
```

Comment:
I wouldn't remove grades here. If the upcoming API call fails the user will see an empty UI. If we leave grades in place at least there will be some data on screen.